seq_strstr/dpu/strstr_main.c

```diff
@@ -0,0 +1,179 @@
+#include <defs.h>
+#include <mram.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <perfcounter.h>
+#include <string.h>
+#include <mutex.h>
+#include "common.h"
+#include <alloc.h>
+#include <built_ins.h>
+#include <defs.h>
+#include <seqread.h>
+
+#include "../../common/include/common.h"
+
+/* dpu marcos */
+#define BLOCK_SIZE 2048 
+#define MAX_BODY_ALLOCTE 4096
+#define MIN_RECORDS_LENGTH 8
+#define STRSTR strstr_org
+
+/* global variables */
+__host uint32_t input_length = 0;
+__host uint32_t output_length = 0;
+__dma_aligned uint8_t key_cache[MAX_KEY_SIZE];
+__host __mram_ptr uint8_t *DPU_BUFFER;
+__host __mram_ptr uint8_t *RECORDS_BUFFER;
+__mram_noinit uint8_t KEY[MAX_KEY_SIZE];
+__host uint32_t input_offset[NR_TASKLETS];
+
+typedef struct in_buffer_context
+{
+	uint8_t *ptr;
+	seqreader_buffer_t cache;
+	seqreader_t sr;
+    __mram_ptr uint8_t *mram_org;
+    uint32_t curr;
+	uint32_t length;
+} in_buffer_context;
+
+// Return values
+typedef enum {
+	STRSTR_JSON_ok = 0, // Success code
+	STRSTR_JSON_INVALID_INPUT,		// Input file has an invalid format
+} strstr_status;
+
+static int find_next_set_bit(unsigned int res, int start) {
+    if(start == 3) {
+        return 4;
+    }
+    
+    for (int i=start; i< 4; i++){
+        if(res & (0x01<<((3-i)*8))){
+            return i;
+        }
+    }
+    return 4;
+}
+
+void shift_same(uint8_t* start, unsigned int *a){
+    *a = 0;
+    *a |= (start[0]<< 24);
+    *a |= (start[0]<< 16);
+    *a |= (start[0]<< 8);
+    *a |= (start[0]);
+}
+
+
+/*
+ * Parse search string from mram
+ */
+bool parseKey() {
+    mram_read(&(KEY[0]), key_cache, MAX_KEY_SIZE);
+    uint8_t* end = memchr(key_cache,'\n' ,MAX_KEY_SIZE);
+
+    if(end != NULL) {
+        *((char*)(end)) = '\0';
+        return true;
+    }
+    else {
+        printf("invalid key\n");
+        return false;
+    }
+}
+
+
+
+
+bool dpu_strstr(struct in_buffer_context *input) {
+    dbg_printf("curr: %u length: %u\n", input->curr, input->length);
+
+    int j = 0; 
+    unsigned int res = 0;
+    unsigned int b = 0;
+    int next =0;
+    
+    do {
+        unsigned int *p_str = seqread_get(input->ptr, sizeof(unsigned int), &input->sr);
+        for(j=0; j< 4; j++) {
+                shift_same(key_cache+j, &b);
+                __builtin_cmpb4_rrr(res, *p_str, b);
+                // jth byte matches
+                if(res & (0x01<<((3-j)*8))){
+                    continue;
+                }
+                else {
+                    next = find_next_set_bit(res, j);
+                    break;
+                }
+            }
+        if (j==4) {
+            return true;
+        }
+        seqread_seek(input->mram_org-4+next, &input->sr);    
```

Comment:
This also needs a comment (why -4 +next)? It somehow seems right, because it's based on "find_next_set_bit" so you are skipping ahead, but I couldn't explain this to someone else if they asked me.

Reply:
Not sure why but the file you are looking is out of date, I should've updated this code a couple of days ago.  In the newer version, this logic has been made much more clear by using a switch statement.